designs/SolutionLoad-CsprojPackageReferenceRestore.md

```diff
@@ -0,0 +1,111 @@
+
+PackageReference is not only supported with SDK projects, but csproj.dll or legacy projects too.
+Certain project types that work with PackageReference wanted a similar experience on solution load, so their project-systems similarly would call a *similar* API that simulates a nomination and thus causing NuGet to run restores on solution load. [See IVsSolutionRestoreService2](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.SolutionRestoreManager.Interop/IVsSolutionRestoreService2.cs).
+
+### Different models for getting reference information in Visual Studio
+
+In PackageReference style projects, there are 2 inputs to restore and that's the NuGet configuration and the project's references, in particular PackageReference and ProjectReference.
+
+While the configuration is calculated in an equivalent way, the approach for getting the package and project references is different:
+
+* SDK-based projects follow a push model (i.e. nomination)
+* csproj.dll projects follow a pull model (i.e. IVsProject4 PackageReference and ProjectReference apis)
+
+### Nomination API Caveat
+
+The nominate API leading to a restore is a leaky abstraction. For SDK based projects, it's not supposed to lead to a restore. It's supposed to ensure that the project is up to date as far as NuGet is concerned. The fact that a restore is run is an implementation detail that ideally wasn't communicated to project-system implementers.
+
+While NuGet knows how to coordinate solution load restore for SDK based projects, it doesn't know how to do the same for these csproj.dll projects and these could lead to a few extra restores.
+While this functionality was added consciously by NuGet, it was not easily obvious what this could lead to.
+
+### Current implementation
+
+In 16.8, we made various improvements to allow us to run only 1 run on solution load. This required changing the batching logic for solution load.
+
+A bit of background around solution load and the activities happening there as of 16.8 P4.
+
+* At solution load, NuGet is able to identify and classify *all* projects in the solution (by listening to the SolutionEvents ProjectAdded).
+* NuGet now knows all projects in the solution that are expected to nominate.
+* When the first nomination comes in, NuGet starts waiting for *all projects* to be nominated.
+* In order to ensure that we don't wait for forever, NuGet has a 10s sliding window. This means that even if all the projects were *not* nominated yet, NuGet will still try to restore as long as 10s since the last nomination have passed.

```

Comment:
Previously I saw issue which try to improve duplicate restore nomination nomination, but couldn't find it anymore.
How about we create tree graph of solution first then instead of waiting 10s can we let in bottom level projects and let them restore since they wouldn't be nominated/restored again. That way we might improve perf for large projects. 
If idea is feasible then you can assign to me to investigate.

Reply:
It's a load statement, but the general idea is something we've considered in the past. 

My best attempt to answer your individual comments. 

> How about we create tree graph of solution first then instead of waiting 10s can we let in bottom level projects and let them restore since they wouldn't be nominated/restored again. 

You are suggesting something close to https://github.com/NuGet/Home/issues/9801. 
The trick here is that we don't know the amount of work necessary to do a partial restore, while it could reduce individual restore times, but it has the potential to lead to slower overall experience. 

Note that in *most* cases, we don't drop out of the sliding window, but we get all nominations instead. 
The complexity of a partial approach is also something that needs to be accounted for, at this point our performance is `good enough` where such a rework might prove to be too costly.